### DISCUSSION TITLE
v2: Local/VS Code version added + major correctness fixes

### DISCUSSION THREAD

**Discussion by edizan12** [edizan12/Synthetic-Intelligence-Architecture#2] [Announcements]
v2: Local/VS Code version added + major correctness fixes

Two things happened in this update:

1. Two ways to run it now. In addition to the original Colab notebook, there's now a plain Python script (main.py) you can run locally in VS Code or any terminal with a CUDA GPU — no notebook, no !pip install, no notebook_login(). Config is passed via CLI flags instead of typed into an input box each run:

bash
pip install transformers torch huggingface_hub bitsandbytes accelerate
huggingface-cli login
python main.py --prompt "Who won the Nobel Prize in Physics in 1823?"

2. Several correctness bugs fixed since the first version. The earliest version of this project had some real problems — worth being upfront about them rather than pretending this was always solid:

Vocabulary alignment was semantically wrong. The original approach compared Llama's and Qwen's logits by truncating both to the same length and comparing them index-by-index. Index 4021 in Llama's vocabulary and index 4021 in Qwen's vocabulary are unrelated tokens — this produced a comparison that looked plausible but wasn't actually meaningful. Fixed by building a real string-based mapping: for each Llama token, check whether its decoded string re-encodes to exactly one Qwen token, and only compare those pairs.
No KV-cache. Every generation step was re-running the full sequence through both models from scratch (model(input_ids) on the whole growing sequence each time) instead of caching past attention states. This is O(n²) — generation got progressively slower the longer the output. Fixed with past_key_values + use_cache=True, so each step only processes the new token.
Raw logit subtraction, uncalibrated. The two models differ in scale and training, so their raw logit magnitudes aren't directly comparable — subtracting them as-is doesn't have a clean interpretation. Fixed by applying log_softmax to both first, so the contrastive step happens in a calibrated probability (log-space) rather than on raw, incomparable magnitudes.
IndexError on Llama-3.2. tokenizer.vocab_size under-reports the model's actual output dimension (it doesn't count all the special/reserved tokens baked into the real logits layer). This caused a shape mismatch crash ([128000] vs [128256]) partway through generation. Fixed by reading the vocab size from model.config.vocab_size instead.
No repetition penalty. Long outputs could get stuck in repetition loops under pure greedy decoding. Added a windowed repetition penalty (looks at the last N generated tokens, not the entire history — penalizing every token ever generated would also suppress normal function words like "the"/"a").
No real benchmarking. Speed and VRAM numbers weren't actually measured before. Now reports real tokens/sec and peak VRAM via torch.cuda.max_memory_allocated() after each generation.

Known limitations that are still open (see README for details): English-only for now (non-English prompts currently produce corrupted output — root cause understood, not yet fixed), greedy decoding only, partial vocabulary coverage between the two tokenizers, and no quantitative hallucination benchmark yet (current examples in the README are qualitative, not a controlled comparison against the expert model alone).

If you cloned or downloaded the notebook before this update, pull the latest or re-clone to get these fixes — the vocabulary-alignment and IndexError issues in particular could affect anyone running the original version against Llama-3.2.
